### ISSUE TITLE
Feature Request: Optionally Keep All Columns in the List Column after `nest`

### ISSUE THREAD

**Issue by billdenney** [tidyverse/tidyr#486]
Could an option to `nest()` be added named something like `retain_all_cols` where the nested data list would have all the columns from the original data?

I'd like to have the option of generating `mydata_nested_with_all_original` directly from `nest`.

``` r
library(tibble)
library(tidyr)
#> Warning: package 'tidyr' was built under R version 3.4.4

mydata <- tibble(foo = rep(1:3, each = 2), bar = 4:9)
mydata_nested <- nest(mydata, -foo)
#> Warning: package 'bindrcpp' was built under R version 3.4.4
mydata_nested$data[[1]]
#> # A tibble: 2 x 1
#>     bar
#>   <int>
#> 1     4
#> 2     5

mydata_nested_with_all_original <- mydata_nested
for (i in seq_len(nrow(mydata_nested_with_all_original))) {
  mydata_nested_with_all_original$data[[i]]$foo <- mydata_nested_with_all_original$foo[[i]]
}
mydata_nested_with_all_original$data[[1]]
#> # A tibble: 2 x 2
#>     bar   foo
#>   <int> <int>
#> 1     4     1
#> 2     5     1
```

With this, a feature would be needed in `unnest()` to confirm that if the column exists in both the list column and the vector columns that the single value in the list column matched the vector column.

**Comment by jennybc** [comment]
Related to #434.

I don't think it's a good idea for a variable `foo`to exist in 2 places in data frame, because it introduces the very real possibility that you'll modify one and not the other and they'll contradict each other. Now what? If it's OK for top-level variable `foo` and the nested variable `foo` to drift apart, then the differently-named copy could be made prior to nesting.

Can you say more about what you're trying to accomplish in terms of workflow?

**Comment by billdenney** [comment]
My use case would be solved by something very similar to that in #434 (and maybe identical).

I understand the desire that `foo` should not exist in two places from a tidy perspective, but from an operations perspective, it is very helpful to have it both places in a few use cases for me.

I have a few different use cases.  Within my PKNCA package, I need to combine different datasets that don't share much structure (data for concentration of medicine in blood and data for the amount of medicine dosed and how it was dosed).  In general, I'd like to be able to do the following (typed directly into GitHub, so there may be bugs but hopefully it communicates the concept):

``` r
calc_conc_dose <- function(conc, dose) {
  # In reality this function is much more complex, but the goal is to have the
  # identifier information seamlessly carried through.
  stop("Invalid Person_Identifier: ", conc$Person_Identifier[1])
}

# Read in the concentration data as d_concentration
# Read in the dosing data as d_dose
d_conc_nested <-
  nest(d_concentration, -Person_Identifier, .key="conc", retain_all_cols=TRUE)
d_dose_nested <-
  nest(d_dose, -Person_Identifier, .key="dose", retain_all_cols=TRUE)
d_conc_dose <- full_join(d_conc_nested, d_dose_nested)
for (idx in seq_len(nrow(d_conc_dose))) {
  calc_conc_dose(d_conc_dose$conc[[idx]], d_conc_dose$dose[[idx]])
}
```

I have a second use case where I often want to make many figures by nesting a data.frame by one variable (e.g. a medical lab test) and then plotting them all separately.  For this one, the suggestion in #434 would work.  With all columns included in the data, I wouldn't have to reach outside of the data for all the information needed to make a figure.  Without the nesting factor inserted, I would require an extra input.

``` r
# Read in my lab data as d_lab
lab_plotter <- function(data) {
  ggplot(data, aes(x=time, y=measurement)) + geom_line() + labs(title=unique(data$Test_Name))
}
d_lab_nested <- nest(d_lab, -Test_Name, retain_all_cols=TRUE)
d_lab_nested$plot <- map(d_lab_nested$data, lab_plotter)
```

**Comment by jennybc** [comment]
In the sketch re: PKNCA, it seems like you don't need the `Person_Identifier` inside `calc_conc_dose()`. You could do a check for valid identifiers after the join and prior to the next computation. And that would be a cleaner design anyway, i.e. `calc_conc_dose()` gets to do exactly what it says on the tin. But I know you said it was just a sketch.

I'm pushing on this because I think needing the same data in two places probably hints at an awkwardness in the workflow. So, I try to publicize the workflows and/or identify missing pieces in the tools.

Shameless self-promotion: fielding lots of good questions like this is what inspired us to do a whole webinar on row-oriented workflows recently. I think it has some good patterns for this sort of work: https://github.com/jennybc/row-oriented-workflows#readme

**Comment by billdenney** [comment]
I've been thinking about this, and I'm coming to the thought that you're probably right:  I should restructure my functions such that they don't use the data in more than one place.

Summary of all of the below:  Without cluttering the argument list, do you know of a way to capture errors, warnings, and messages which provides the actual code location of the source of the error-warning-message?

To add some detail, one complexity of the work that PKNCA does which makes it infeasible to just use `summarize()` on the row is that calculation results may depend on each other.  So, if I'm requesting a calculation of parameter `C` it may depend on `A` and `B` which are automatically calculated first within PKNCA (and there are up to about 4 layers of potential dependencies of calculations).

To illustrate the workflow as it currently is in a bit more detail, there are functions that do calculations on the data `calc_conc_dose()` that may operate either at the dataset level (multiple individuals), the individual level (one or more calculations), or the calculation level (return a single result).  My goal had been to capture all the warnings and errors at the deepest level (the calculation level) and keep the outer levels simpler to just pass in the data and an indicator of what the current context is (individual and calculation number).  With that, it's simpler to pass in the data and a consistent set of identifiers (the columns in the data that define the individual and calculation number).

To work on a row basis, I'll need to either run the deeper levels (individual and calculation) within a combination of `purrr::safely()` and `purrr::quietly()` or have them modify their output as though they were running within those `purrr` functions (adding error, warning, and message attributes to the output).  I see merit to that, but it does make the inner functions a bit more complex in how they must perform error, warning, and message handling.

As a secondary benefit, reorganizing my code like this would work around this base R bug/feature request: https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17122 .

As a negative, the actual location of the warning or error would be lost since it would be raised from a different code location than the actual error (maybe there is a fancier way around this than I'm aware; `purrr::silently()` loses the actual error function, changing it to `.f()` as far as I can tell, see below).

``` r
library(purrr)

function_inner <- function() {
  stop("The error is in function_inner")
}

function_outer <- function() {
  safely(function_inner)()
}

foo <- function_outer()
dput(foo)
#> structure(list(result = NULL, error = structure(list(message = "The error is in function_inner", 
#>     call = .f(...)), .Names = c("message", "call"), class = c("simpleError", 
#> "error", "condition"))), .Names = c("result", "error"))
```

P.S.  I'm always a fan of self-promotion with a disclaimer!

**Comment by jennybc** [comment]
I confess I'm probably not following the details re: your desires to capture (and hold on to?) errors/warnings/messages. We've probably crossed into a design topic that's better to raise in https://community.rstudio.com/c/package-development.

FWIW the tidyverse/r-lib team is going more down the road of writing custom conditions: https://adv-r.hadley.nz/conditions.html#custom-conditions

**Comment by billdenney** [comment]
I probably wasn’t fully clear.

For this issue, you’ve convinced me of a better way to do things with `nest()`.

For error handling, I think that my general solution will be to capture and increase the attributes of the warnings and errors moving up the call stack.  (I’ll move that discussion elsewhere because it doesn’t belong here.)

**Comment by mikegilchrist** [comment]
I came across this because I wanted to do the same thing, retain the original column used with `nest()`, setting concerns about having data in 2 places aside (which since this is a reference table, I'm not worried about).  I came up with a simple work around and I thought I'd share it here.  It works by duplicating the nested column and then unnesting that second column.  

`
data %>% 
   nest( nested_a = a) %>% 
   mutate(tmp = nested_a) %>%
   unnest(tmp)
`
